API Unusable - Model `did` was null

Had a user report: API was no longer working.

<details><summary>Logcat</summary>

```
06-06 01:01:14.253 17345 17390 E AnkiDroid: CardContentProvider/ Error parsing JSONArray
06-06 01:01:14.253 17345 17390 E AnkiDroid: com.ichi2.utils.JSONException: org.json.JSONException: Value null at did of type org.json.JSONObject$1 cannot be converted to long
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at com.ichi2.utils.JSONObject.getLong(JSONObject.java:2)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at com.ichi2.anki.provider.CardContentProvider.addModelToCursor(CardContentProvider.java:19)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at com.ichi2.anki.provider.CardContentProvider.query(CardContentProvider.java:84)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at android.content.ContentProvider.query(ContentProvider.java:1276)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at android.content.ContentProvider.query(ContentProvider.java:1369)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at android.content.ContentProvider$Transport.query(ContentProvider.java:275)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:114)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at android.os.Binder.execTransactInternal(Binder.java:1032)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at android.os.Binder.execTransact(Binder.java:1005)
06-06 01:01:14.253 17345 17390 E AnkiDroid: Caused by: org.json.JSONException: Value null at did of type org.json.JSONObject$1 cannot be converted to long
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at org.json.JSON.typeMismatch(JSON.java:101)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at org.json.JSONObject.getLong(JSONObject.java:527)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at com.ichi2.utils.JSONObject.getLong(JSONObject.java:1)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     ... 8 more
06-06 01:01:14.253 17345 17390 E AnkiDroid: 
06-06 01:01:14.253 17345 17390 E AnkiDroid: com.ichi2.utils.JSONException: org.json.JSONException: Value null at did of type org.json.JSONObject$1 cannot be converted to long
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at com.ichi2.utils.JSONObject.getLong(JSONObject.java:2)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at com.ichi2.anki.provider.CardContentProvider.addModelToCursor(CardContentProvider.java:19)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at com.ichi2.anki.provider.CardContentProvider.query(CardContentProvider.java:84)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at android.content.ContentProvider.query(ContentProvider.java:1276)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at android.content.ContentProvider.query(ContentProvider.java:1369)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at android.content.ContentProvider$Transport.query(ContentProvider.java:275)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:114)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at android.os.Binder.execTransactInternal(Binder.java:1032)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at android.os.Binder.execTransact(Binder.java:1005)
06-06 01:01:14.253 17345 17390 E AnkiDroid: Caused by: org.json.JSONException: Value null at did of type org.json.JSONObject$1 cannot be converted to long
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at org.json.JSON.typeMismatch(JSON.java:101)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at org.json.JSONObject.getLong(JSONObject.java:527)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     at com.ichi2.utils.JSONObject.getLong(JSONObject.java:1)
06-06 01:01:14.253 17345 17390 E AnkiDroid:     ... 8 more
06-06 01:01:14.253 17345 17390 E ContentProviderNative: onTransact error from {P:14524;U:10105}
06-06 01:01:14.253 17345 17390 E DatabaseUtils: Writing exception to parcel
06-06 01:01:14.253 17345 17390 E DatabaseUtils: java.lang.IllegalArgumentException: Model 1342706442510 is malformed
06-06 01:01:14.253 17345 17390 E DatabaseUtils:         at com.ichi2.anki.provider.CardContentProvider.addModelToCursor(CardContentProvider.java:32)
06-06 01:01:14.253 17345 17390 E DatabaseUtils:         at com.ichi2.anki.provider.CardContentProvider.query(CardContentProvider.java:84)
06-06 01:01:14.253 17345 17390 E DatabaseUtils:         at android.content.ContentProvider.query(ContentProvider.java:1276)
06-06 01:01:14.253 17345 17390 E DatabaseUtils:         at android.content.ContentProvider.query(ContentProvider.java:1369)
06-06 01:01:14.253 17345 17390 E DatabaseUtils:         at android.content.ContentProvider$Transport.query(ContentProvider.java:275)
06-06 01:01:14.253 17345 17390 E DatabaseUtils:         at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:114)
06-06 01:01:14.253 17345 17390 E DatabaseUtils:         at android.os.Binder.execTransactInternal(Binder.java:1032)
06-06 01:01:14.253 17345 17390 E DatabaseUtils:         at android.os.Binder.execTransact(Binder.java:1005)
06-06 01:01:14.253 17345 17390 E DatabaseUtils: Caused by: com.ichi2.utils.JSONException: org.json.JSONException: Value null at did of type org.json.JSONObject$1 cannot be converted to long
06-06 01:01:14.253 17345 17390 E DatabaseUtils:         at com.ichi2.utils.JSONObject.getLong(JSONObject.java:2)
06-06 01:01:14.253 17345 17390 E DatabaseUtils:         at com.ichi2.anki.provider.CardContentProvider.addModelToCursor(CardContentProvider.java:19)
06-06 01:01:14.253 17345 17390 E DatabaseUtils:         ... 7 more
06-06 01:01:14.253 17345 17390 E DatabaseUtils: Caused by: org.json.JSONException: Value null at did of type org.json.JSONObject$1 cannot be converted to long
06-06 01:01:14.253 17345 17390 E DatabaseUtils:         at org.json.JSON.typeMismatch(JSON.java:101)
06-06 01:01:14.253 17345 17390 E DatabaseUtils:         at org.json.JSONObject.getLong(JSONObject.java:527)
06-06 01:01:14.253 17345 17390 E DatabaseUtils:         at com.ichi2.utils.JSONObject.getLong(JSONObject.java:1)
06-06 01:01:14.253 17345 17390 E DatabaseUtils:         ... 8 more
```
</details>

The cause was a model with a `null` did. Confirmed via inspection from the Anki Desktop Python console.

This could be an API issue on our end, or corruption on the Anki Desktop side.

We have one instance where we can store `null` for the `did` of a model (a string is used, rather than a long):

https://github.com/ankidroid/Anki-Android/blob/1e4953b7de30c603c2564623775d3a42632778ab/AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java#L545

We can fix this by providing a reasonable default value

--- a/AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/ContentProviderTest.java
+++ b/AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/ContentProviderTest.java
@@ -964,7 +964,23 @@ public class ContentProviderTest {
 
     }
 
-    private Collection reopenCol() {
+
+
+    /** Test that a null did will not crash the provider (#6378) */
+     @Test
+     public void testProviderProvidesDefaultForEmptyModelDeck() {
+         Collection col = getCol();
+         col.getModels().all().get(0).put("did", JSONObject.NULL);
+         col.save();
+
+         final ContentResolver cr = InstrumentationRegistry.getInstrumentation().getTargetContext().getContentResolver();
+         // Query all available models
+         final Cursor allModels = cr.query(FlashCardsContract.Model.CONTENT_URI, null, null, null, null);
+         assertNotNull(allModels);
+     }
+
+
+ private Collection reopenCol() {
         CollectionHelper.getInstance().closeCollection(false, "ContentProviderTest: reopenCol");
         return getCol();
     }

--- a/AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java
@@ -42,6 +42,7 @@ import com.ichi2.anki.FlashCardsContract;
 import com.ichi2.anki.FlashCardsContract.CardTemplate;
 import com.ichi2.anki.exception.ConfirmModSchemaException;
 import com.ichi2.compat.CompatHelper;
+import com.ichi2.libanki.Consts;
 import com.ichi2.libanki.sched.AbstractSched;
 import com.ichi2.libanki.Card;
 import com.ichi2.libanki.Collection;
@@ -1083,7 +1084,8 @@ public class CardContentProvider extends ContentProvider {
                 } else if (column.equals(FlashCardsContract.Model.CSS)) {
                     rb.add(jsonObject.getString("css"));
                 } else if (column.equals(FlashCardsContract.Model.DECK_ID)) {
-                    rb.add(jsonObject.getLong("did"));
+                    //#6378 - Anki Desktop changed schema temporarily to allow null
+                    rb.add(jsonObject.optLong("did", Consts.DEFAULT_DECK_ID));
                 } else if (column.equals(FlashCardsContract.Model.SORT_FIELD_INDEX)) {
                     rb.add(jsonObject.getLong("sortf"));
                 } else if (column.equals(FlashCardsContract.Model.TYPE)) {

